File: apps/java/src/main/java/tdd/fizzbuzz/domain/model/FizzBuzzList.java
package tdd.fizzbuzz.domain.model;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.function.Function;
import java.util.stream.Collectors;

public final class FizzBuzzList {

    private final List<FizzBuzzValue> values;

    public FizzBuzzList(List<FizzBuzzValue> values) {
        this.values = new ArrayList<>(values);
    }

    public List<FizzBuzzValue> getValues() {
        return new ArrayList<>(values);
    }

    public FizzBuzzValue get(int index) {
        return values.get(index);
    }

    public int size() {
        return values.size();
    }

    public FizzBuzzList add(List<FizzBuzzValue> newValues) {
        List<FizzBuzzValue> combined = new ArrayList<>(values);
        combined.addAll(newValues);
        return new FizzBuzzList(combined);
    }

    public FizzBuzzList filter(Predicate<FizzBuzzValue> predicate) {
        List<FizzBuzzValue> filtered = values.stream()
                .filter(predicate)
                .collect(Collectors.toList());
        return new FizzBuzzList(filtered);
    }

    public Optional<FizzBuzzValue> findFirst(Predicate<FizzBuzzValue> predicate) {
        return values.stream()
                .filter(predicate)
                .findFirst();
    }

    public <R> List<R> map(Function<FizzBuzzValue, R> mapper) {
        return values.stream()
                .map(mapper)
                .collect(Collectors.toList());
    }

    public List<String> toStringValues() {
        return values.stream()
                .map(FizzBuzzValue::getValue)
                .collect(Collectors.toList());
    }

    public Map<String, List<FizzBuzzValue>> groupByValue() {
        return values.stream()
                .collect(Collectors.groupingBy(
                        FizzBuzzValue::getValue,
                        LinkedHashMap::new,
                        Collectors.toList()));
    }

    public Map<String, Long> countByValue() {
        return values.stream()
                .collect(Collectors.groupingBy(
                        FizzBuzzValue::getValue,
                        LinkedHashMap::new,
                        Collectors.counting()));
    }

    public String joining(String delimiter) {
        return values.stream()
                .map(FizzBuzzValue::getValue)
                .collect(Collectors.joining(delimiter));
    }

    @Override
    public String toString() {
        return values.toString();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null || getClass() != obj.getClass()) {
            return false;
        }
        FizzBuzzList that = (FizzBuzzList) obj;
        return Objects.equals(values, that.values);
    }

    @Override
    public int hashCode() {
        return Objects.hash(values);
    }
}

File: apps/dotnet/FizzBuzz/Domain/Model/FizzBuzzList.cs
namespace FizzBuzz.Domain.Model;

public sealed class FizzBuzzList : IEquatable<FizzBuzzList>
{
    private readonly List<FizzBuzzValue> _values;

    public IReadOnlyList<FizzBuzzValue> Values => _values.AsReadOnly();

    public FizzBuzzList()
    {
        _values = new List<FizzBuzzValue>();
    }

    public FizzBuzzList(List<FizzBuzzValue> values)
    {
        _values = new List<FizzBuzzValue>(values);
    }

    public FizzBuzzValue Get(int index) => _values[index];

    public int Count => _values.Count;

    public FizzBuzzList Add(FizzBuzzValue value)
    {
        var newValues = new List<FizzBuzzValue>(_values) { value };
        return new FizzBuzzList(newValues);
    }

    public FizzBuzzList AddRange(IEnumerable<FizzBuzzValue> values)
    {
        var newValues = new List<FizzBuzzValue>(_values);
        newValues.AddRange(values);
        return new FizzBuzzList(newValues);
    }

    public FizzBuzzList Filter(Func<FizzBuzzValue, bool> predicate)
    {
        return new FizzBuzzList(_values.Where(predicate).ToList());
    }

    public FizzBuzzValue? FindFirst(Func<FizzBuzzValue, bool> predicate)
    {
        return _values.FirstOrDefault(predicate);
    }

    public List<string> ToStringValues()
    {
        return _values.Select(v => v.Value).ToList();
    }

    public Dictionary<string, int> CountByValue()
    {
        return _values
            .GroupBy(v => v.Value)
            .ToDictionary(g => g.Key, g => g.Count());
    }

    public override string ToString()
    {
        return string.Join(", ", _values.Select(v => v.ToString()));
    }

    public bool Equals(FizzBuzzList? other)
    {
        if (other is null) return false;
        if (ReferenceEquals(this, other)) return true;
        return _values.SequenceEqual(other._values);
    }

    public override bool Equals(object? obj) => Equals(obj as FizzBuzzList);

    public override int GetHashCode()
    {
        var hash = new HashCode();
        foreach (var value in _values)
            hash.Add(value);
        return hash.ToHashCode();
    }
}
